pkg/liquidity-mining/contracts/VotingEscrowRemapper.sol

```diff
@@ -0,0 +1,306 @@
+        return _localToRemoteAddressMap[chainId][localUser];
+    }
+
+    function getRemappingManager(address localUser) public view returns (address) {
+        return _localRemappingManager[localUser];
+    }
+
+    function setOmniVotingEscrow(IOmniVotingEscrow omniVotingEscrow) external authenticate {
+        _omniVotingEscrow = omniVotingEscrow;
+        emit OmniVotingEscrowUpdated(omniVotingEscrow);
+    }
+
+    // Remapping Setters
+
+    /**
+     * @notice Sets up a mapping from `localUser`'s veBAL balance to `remoteUser` for chain `chainId`.
+     * @dev In order to set up a remapping on this contract, `localUser` must be a smart contract which has been
+     * allowlisted to hold veBAL. EOAs are expected to set up any delegation of their veBAL on L2s directly.
+     * @param localUser - The address of a contract allowlisted on the `SmartWalletChecker`.
+     * @param remoteUser - The address to receive `localUser`'s balance of veBAL on the remote chain.
+     * @param chainId - The chain id of the remote chain on which `remoteUser` resides.
+     */
+    function setNetworkRemapping(
+        address localUser,
+        address remoteUser,
+        uint256 chainId
+    ) external payable nonReentrant {
+        _require(msg.sender == localUser || msg.sender == _localRemappingManager[localUser], Errors.SENDER_NOT_ALLOWED);
+        require(_isAllowedContract(localUser), "Only contracts which can hold veBAL can set up a mapping");
+        require(remoteUser != address(0), "Zero address cannot be used as remote user");
+        IOmniVotingEscrow omniVotingEscrow = getOmniVotingEscrow();
+        require(address(omniVotingEscrow) != address(0), "Omni voting escrow not set");
+
+        // We keep a 1-to-1 local-remote mapping for each chain.
+        // If A --> B (i.e. A in the local chain is remapped to B in the remote chain), to keep the state consistent
+        // the user effectively 'owns' both A and B in both chains.
+        //
+        // This means that whenever a new address is created (assuming A --> B previously):
+        // - The remote address must not already be in use by another local user (C --> B is forbidden).
+        // - The remote address must not be a local address that has already been remapped (C --> A is forbidden).
+        // - The local address must not be the target remote address for another local user (B --> C is forbidden).
+        //
+        // Note that this means that it is possible to frontrun this call to grief a user by taking up their
+        // selected remote address before they do so. This is mitigated somewhat by restricting potential attackers to
+        // the set of contracts that are allowlisted to hold veBAL (and their remapping managers). Should
+        // one of them grief, then Balancer governance can remove them from these allowlists.
+
+        // Assuming A --> B, then C cannot be remapped to B.
```

Comment:
```suggestion
        // B cannot be remapped to (i.e. be a remote) if a prior A --> B mapping exists.
```

Reply:
I found the original phrasing very confusing (spent over 10 minutes re-reading these statements). I think this new way of stating them is simpler.